tokio/src/sync/notify.rs

```diff
@@ -213,24 +213,17 @@ pub struct Notify {
     waiters: Mutex<WaitList>,
 }
 
-#[derive(Debug, Clone, Copy)]
-enum NotificationType {
-    // Notification triggered by calling `notify_waiters`
-    AllWaiters,
-    // Notification triggered by calling `notify_one`
-    OneWaiter,
-}
-
 #[derive(Debug)]
 struct Waiter {
     /// Intrusive linked-list pointers.
     pointers: linked_list::Pointers<Waiter>,
 
     /// Waiting task's waker.
-    waker: Option<Waker>,
+    waker: UnsafeCell<Option<Waker>>,
 
-    /// `true` if the notification has been assigned to this waiter.
-    notified: Option<NotificationType>,
+    /// Notification for this waiter. It is an enum and uses
+    /// `NOTIFICATION_{NONE, ONE, ALL}` values.
+    notification: AtomicUsize,
```

Comment:
I think the code is ok, but the safety invariants invovled with this atomic should be documented more clearly.